openfl-tutorials/interactive_api/HPU/PyTorch_MedMNIST_2D/workspace/HPU_Pytorch_MedMNIST_2D.ipynb

```diff
@@ -0,0 +1,627 @@
+   "execution_count": null,
+   "id": "f2154486",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8d1c78ee",
+   "metadata": {},
+   "source": [
+    "### Register model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "59831bcd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from copy import deepcopy\n",
+    "\n",
+    "framework_adapter = 'openfl.plugins.frameworks_adapters.pytorch_adapter.FrameworkAdapterPlugin'\n",
+    "MI = ModelInterface(model=model, optimizer=optimizer, framework_plugin=framework_adapter)\n",
+    "\n",
+    "# Save the initial model state\n",
+    "initial_model = deepcopy(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "849c165b",
+   "metadata": {},
+   "source": [
+    "## Define and register FL tasks"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4ff463bd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TI = TaskInterface()\n",
+    "\n",
+    "train_custom_params={'criterion':criterion,'task':task, 'device':'hpu'}\n",
+    "\n",
+    "# Task interface currently supports only standalone functions.\n",
+    "@TI.add_kwargs(**train_custom_params)\n",
+    "@TI.register_fl_task(model='model', data_loader='train_loader',\n",
+    "                     device='device', optimizer='optimizer')\n",
+    "def train(model, train_loader, device, optimizer, criterion, task):\n",
+    "    total_loss = []\n",
+    "    \n",
+    "    train_loader = tqdm.tqdm(train_loader, desc=\"train\")\n",
+    "    model.train()\n",
+    "    model.to(device)\n",
+    "    \n",
+    "    for inputs, targets in train_loader:\n",
+    "    \n",
+    "        optimizer.zero_grad()\n",
+    "        outputs = model(inputs.to(device))\n",
+    "        \n",
+    "        if task == 'multi-label, binary-class':\n",
+    "            targets = targets.to(torch.float32).to(device)\n",
+    "            loss = criterion(outputs, targets)\n",
+    "        else:\n",
+    "            targets = torch.squeeze(targets, 1).long().to(device)\n",
+    "            loss = criterion(outputs, targets)\n",
+    "        \n",
+    "        total_loss.append(loss.item())\n",
+    "        \n",
+    "        loss.backward()\n",
+    "        htcore.mark_step()\n",
+    "        optimizer.step()\n",
+    "        htcore.mark_step()\n",
+    "\n",
+    "    return {'train_loss': np.mean(total_loss),}\n",
+    "\n",
+    "\n",
+    "\n",
+    "val_custom_params={'criterion':criterion, 'task':task, 'device':'hpu'}\n",
+    "\n",
+    "@TI.add_kwargs(**val_custom_params)\n",
+    "@TI.register_fl_task(model='model', data_loader='val_loader', device='device')\n",
+    "def validate(model, val_loader, device, criterion, task):\n",
+    "\n",
+    "    val_loader = tqdm.tqdm(val_loader, desc=\"validate\")\n",
+    "    model.eval()\n",
+    "    model.to(device)\n",
```

Comment:
What is the device name for HPU? It is not obvious how tensors end up in HPU memory

Reply:
i] To get the details about HPU device like HPU device name, index number we need to import below package:
import habana_frameworks.torch.hpu as hthpu

The output of the below API's for current example is highlighted below:

hthpu.get_device_name() - Returns the name of the HPU device --> **'GAUDI'**
hthpu.current_device() - Returns the index of the current selected HPU device --> **0**

ii] First we initialize `device = 'hpu'` , then move Model and Tensors to device('HPU') memory using `model.to(device)` , `input.to(device)`, `targets.to(device)` etc

We can verify this by checking the device that the tensor is present in before and after moving to HPU device by using `tensor.device`, ex: `input.device`, `target.device` . 

o/p before moving to HPU: `device(type='cpu')`
o/p after moving to HPU: `device(type='hpu', index=0)`

Reply:
But the `device` variable will be set to `cpu`  in this method, this is why I ask

Reply:
We have set `device='hpu'` in custom_params before moving model or tensors to device

`train_custom_params={'criterion':criterion,'task':task, 'device':'hpu'}`
`val_custom_params={'criterion':criterion, 'task':task, 'device':'hpu'}`

Reply:
That is an interesting case of overwriting the task contract parameters. It is not obvious which one will be passed in runtime. It may be better to hardcode the device inside the task code.

Reply:
The parameters passed in [add_kwargs()](https://openfl.readthedocs.io/en/latest/running_the_federation.html#id4:~:text=%40TI.-,add_kwargs,-(**task_settings) ) overrides the local variables, and the device passed as HPU here gets the precedence and it persists. For the current example when I print the location of the tensors, variables and model parameters in both train and validation task, it shows that they reside on HPU as shown below:

**Train Task:**
![image](https://user-images.githubusercontent.com/73102565/212242122-c02fd94c-2737-49f8-9631-cd2cf2ad201f.png)

**Validation Task**
![image](https://user-images.githubusercontent.com/73102565/212242299-594b697b-08d8-4ce5-8614-b0afc4e231ab.png)

Can you help me to know if I am missing something.

Reply:
I understand that it works, but you should not rely on this behavior, I mean overriding parameters.
Moreover, you probably want to emphasize that the device used for training is HPU, so I still think it is better to put
`device='HPU:0` in both tasks' code.

Reply:
Got it, sure will make that change.

Reply:
Please make a comment there, something like:
```
# This task is designed to run on remote nodes equipped with HPU
# so we hardcode the device name here.
```

Reply:
Done